enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,689 @@
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: RouteAdvertisements
+metadata:
+  name: default
+spec:
+  targetVRF: auto 
+  advertisements:
+    podNetwork: true
+  networkSelector:
+    matchExpressions:
+      - { key: k8s.ovn.org/metadata.name, operator: In, values: [blue,red] } 
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/hostname: ovn-worker
+  frrConfigurationSelector:
+    matchLabels:
+      routeAdvertisements: vpn-blue-red
+```
+
+In the above CR, the targetVRF is set to auto, meaning the advertisements will occur within the VRF corresponding to the
+individual networks selected. In this case, the pod subnet for blue will be advertised over the blue VRF, while the pod
+subnet for red will be advertised over the red VRF. OVN-Kubernetes creates the following FRRConfiguration:
+
+```yaml
+apiVersion: frrk8s.metallb.io/v1beta1
+kind: FRRConfiguration
+metadata:
+  name: route-advertisements-blue
+  namespace: metallb-system
+spec:
+  bgp:
+    routers:
+    - asn: 64512
+      neighbors:
+      - address: 182.18.0.5
+        asn: 64512
+        toAdvertise:
+          allowed:
+            prefixes:
+               - 10.0.1.0/24
+      vrf: blue
+      prefixes:
+        - 10.0.1.0/24
+    - asn: 64512
+      neighbors:
+      - address: 192.18.0.5
+        asn: 64512
+        toAdvertise:
+          allowed:
+            prefixes:
+            - 10.0.1.0/24
+      vrf: red
+      prefixes:
+         - 10.0.1.0/24  
+  nodeSelector:
+     matchLabels:
+        kubernetes.io/hostname: ovn-worker
+```
+
+OVN-Kubernetes uses the configuration already in place in FRR and the desired RouteAdvertisements to generate the above
+FRRConfiguration. For filtering or choosing what routes to receive, a user should do that in the FRRConfiguration they
+declare for peering.
+
+Note, VRF-Lite is only available when using Local Gateway Mode (routingViaHost: true) in OVN-Kubernetes.
+
+#### Feature Compatibility
+
+##### Multiple External Gateways (MEG)
+
+When using BGP to learn routes to next hops, there can be overlap with gateways detected by the MEG feature. MEG may
+still be configured along with BFD in OVN, and overlapping routes learned from BGP will be ignored by OVN-Kubernetes.
+BFD may also be configured in FRR, for immediate purging of learned routes by FRR.
+
+A user can also configure RouteAdvertisements for namespaces affected by MEG. Since MEG directly uses pod IPs
+(in disable-snat-multiple-gws mode), the external gateway needs to know where to route pod IPs for ingress and egress
+reply traffic. Traditionally this is done by the gateway having its own operator to detect pod subnets via kubernetes
+API. With BGP, this is no longer necessary. A user can simply configure RouteAdvertisements, and the pod subnet routes
+will be dynamically learned by an external gateway capable of BGP peering.
+
+##### Egress IP
+
+EgressIP feature that is dynamically moved between nodes. By enabling the feature in RouteAdvertisements, OVN-Kubernetes
```

Comment:
Can you provide RouteAdvertisements example here for both primary EIP and secondary EIP scenarios ?
For secondary EIP, it may have to choose another secondary interface on the node for establishing BGP session with a neighbor. doesn't it require changes to CR to accommodate source address or interface ?

Reply:
originally we planned on providing egress IP as well as other features on secondary networks. it looks like that functionality is no longer needed now that we can provide a layer 2 type network as the primary network. So we don't need to worry about providing egress IP functionality on secondary networks.

For providing egress ip multi-nic functionality (which I think is really what you are getting at) across mutliple interfaces for different user defined networks there are 2 use cases.

1. Single BGP instance operating in the default VRF with user defined networks leaking into it (targetVRF=default):
 The user would be responsible for setting up BGP peering over those other NICs. Egress IP reachability would be advertised out of all of them. But egress traffic would only go out one of the interfaces.
 2. Multiple VRF BGP instances, using VRF lite to carry the user defined networks over specific interfaces (targetVRF=auto):
 The egress IP would only be advertised over the VRF instance that it maps to, and subsequently the host interface where the BGP peering is happening for that VRF. If egress IP multi-nic was configured to use an interface other than this VRF interface, it should be considered a misconfiguration. We can choose to explicitly raise an error here and/or forbid this behavior.

Does that makes sense?

Reply:
ok @trozet , for case 1, doesn't this just pick up primary interface ip for bgp peering (the same ip is used as next hop for pod ip, egressip routes), isn't it ? that's why i thought of providing an option to choose required interface for bgp peering.

Reply:
Hmm so you are thinking of checking to see if BGP is enabled on this network, and its advertising egress IP, then there should be a BGP peer on the selected egress IP interface? If so, it might be useful to warn the user or maybe post an event to the egress IP, but I'm not sure we want to forbid it.